database/merkledb/forest_persistence.go

```diff
@@ -0,0 +1,166 @@
+package merkledb
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/database/versiondb"
+
+	"github.com/ava-labs/avalanchego/codec/linearcodec"
+
+	"github.com/ava-labs/avalanchego/codec"
+	"github.com/ava-labs/avalanchego/database"
+)
+
+// ForestPersistence holds the DB + the RootNode
+type ForestPersistence struct {
+	db    database.Database
+	codec codec.Manager
+}
+
+// NewForestPersistence creates a new Persistence
+func NewForestPersistence(db database.Database) (Persistence, error) {
+	c := linearcodec.NewDefault()
+	_ = c.RegisterType(&BranchNode{})
+	_ = c.RegisterType(&LeafNode{})
+	_ = c.RegisterType(&RootNode{})
+	codecManager := codec.NewDefaultManager()
+	if err := codecManager.RegisterCodec(0, c); err != nil {
+		return nil, err
+	}
+
+	persistence := ForestPersistence{
+		db:    versiondb.New(db),
+		codec: codecManager,
+	}
+
+	return &persistence, nil
+}
+
+// GetNodeByUnitKey fetches a Node given a StorageKey
+func (fp *ForestPersistence) GetNodeByHash(nodeHash []byte) (Node, error) {
+	nodeBytes, err := fp.db.Get(nodeHash)
+	if err != nil {
+		return nil, err
+	}
+
+	var node Node
+	_, err = fp.codec.Unmarshal(nodeBytes, &node)
+	node.SetPersistence(fp)
```

Comment:
If `Unmarshal` returns an error, it is not safe to call `node.SetPersistence(fp)`.